### ISSUE TITLE
Where do we get the result file for all evaluation

### ISSUE THREAD

**Issue by DhavalRepo18** [TheDatumOrg/TSB-AD#19]
Assuming we do not want to run the benchmark, but want to run a new method and get the comparison done. What is the process?

**Comment by qhliu26** [comment]
Hi! You can now access the evaluation results on the benchmark here https://github.com/TheDatumOrg/TSB-AD/tree/main/benchmark_exp/benchmark_eval_results. Let us know if any help is needed.

**Comment by DhavalRepo18** [comment]
Thanks.

**Comment by DhavalRepo18** [comment]
@qhliu26 thanks for uploading the (subset) of results. I have one question what if results are not reproducible? Do we need further investigation? or all care has been taken to ensure results are reproducible?

**Comment by qhliu26** [comment]
Hi! Based on my experience, various factors such as hyperparameter settings, normalization techniques, random seeds, or even different Python versions can impact the performance and evaluation results to some extent. Some of these variations can be mitigated by running the algorithms multiple times and reporting the results with error bars. Additionally, comparing performance across the entire benchmark can provide a more reliable and robust evaluation. Specifically the evaluation results currently available on github are produced under numpy random seed 2024 and python 3.9. If there are specific algorithms for which you have observed significant deviations from our reported results, please let us know. We will be happy to investigate further.

**Comment by User-Pass320** [comment]
Hello dear qhliu26, I configured the environment and installed the dependent packages based on the readme file, and executed the following commands using the built-in test data. 

![Image](https://github.com/user-attachments/assets/89cdad84-3574-475c-b106-24fe96f73a80)

![Image](https://github.com/user-attachments/assets/99428a13-e787-467a-a40c-5ee11c2227c6)

Why is there such a big difference in the results?

**Comment by qhliu26** [comment]
Hi, thank you for reproducing the results! We had a recent update on the dataloader class, which creates such a discrepancy. For a quick fix on this and to reproduce the results on the benchmark, you can replace the ForecastDataset class (TSB-AD/TSB_AD/utils/dataset.py) with the following code (previous version). We will fix the code ASAP!

class ForecastDataset(torch.utils.data.Dataset):
    def __init__(self, data, window_size, pred_len, normalize=True):
        super().__init__()
        self.normalize = normalize

        if self.normalize:
            data_mean = np.mean(data, axis=0)
            data_std = np.std(data, axis=0)
            data_std = np.where(data_std == 0, epsilon, data_std)
            self.data = (data - data_mean) / data_std
        else:
            self.data = data

        self.window_size = window_size
        
        if data.shape[1] == 1:
            data = data.squeeze()
            self.len, = data.shape
            self.sample_num = max(self.len - self.window_size - pred_len + 1, 0)
            X = torch.zeros((self.sample_num, self.window_size))
            Y = torch.zeros((self.sample_num, pred_len))
            
            for i in range(self.sample_num):
                X[i, :] = torch.from_numpy(data[i : i + self.window_size])
                Y[i, :] = torch.from_numpy(np.array(
                    data[i + self.window_size: i + self.window_size + pred_len]
                ))
            
            self.samples, self.targets = torch.unsqueeze(X, -1), torch.unsqueeze(Y, -1)
    
        else:
            self.len = self.data.shape[0]
            self.sample_num = max(self.len - self.window_size - pred_len + 1, 0)

            X = torch.zeros((self.sample_num, self.window_size, self.data.shape[1]))
            Y = torch.zeros((self.sample_num, pred_len, self.data.shape[1]))

            for i in range(self.sample_num):
                X[i, :] = torch.from_numpy(data[i : i + self.window_size, :])
                Y[i, :] = torch.from_numpy(data[i + self.window_size: i + self.window_size + pred_len, :])
            
            self.samples, self.targets = X, Y

    def __len__(self):
        return self.sample_num

    def __getitem__(self, index):
        return self.samples[index, :, :], self.targets[index, :, :]

**Comment by qhliu26** [comment]
Overall, we would recommend using the latest version of the code, which provides a more robust and reliable data loading process. We will update the benchmark results later accordingly!

**Comment by User-Pass320** [comment]
Hello qhliu26, thank you for your quick reply. When I adjusted the code according to your modification suggestions, the result did improve. The result of the re-execution is as follows:

![Image](https://github.com/user-attachments/assets/fa469baf-809b-4dd1-91d6-51d2811b9370)

But the current result of OmniAnomaly is much higher than the benchmark given(0.028071), while CNN is the opposite(0.414392). I wonder if this result is within reasonable expectations?

**Comment by qhliu26** [comment]
OmniAnomaly uses `ReconstructDataset` in the dataloading step. Here is the previous version:

class ReconstructDataset(torch.utils.data.Dataset):

    def __init__(self, data, window_size, stride=1, normalize=True):
        super().__init__()
        self.normalize = normalize

        if self.normalize:
            data_mean = np.mean(data, axis=0)
            data_std = np.std(data, axis=0)
            data_std = np.where(data_std == 0, epsilon, data_std)
            self.data = (data - data_mean) / data_std
        else:
            self.data = data

        self.window_size = window_size
        self.stride = stride
        
        if data.shape[1] == 1:
            data = data.squeeze()
            self.len, = data.shape
            self.sample_num = max(0, (self.len - self.window_size) // self.stride + 1)
            
            X = torch.zeros((self.sample_num, self.window_size))
            
            for i in range(self.sample_num):
                X[i, :] = torch.from_numpy(data[i * stride : i * stride + self.window_size])
                
            self.samples, self.targets = torch.unsqueeze(X, -1), torch.unsqueeze(X, -1)
        else:
            self.sample_num = data.shape[0]

            X = torch.zeros((data.shape[0], data.shape[1]))            
            for i in range(self.sample_num):
                X[i, :] = torch.from_numpy(data[i,:])

            self.samples, self.targets = X, X            

    def __len__(self):
        if self.data.shape[1] == 1:
            return self.sample_num
        else:
            return self.data.shape[0]
    
    def __getitem__(self, index):
        if self.data.shape[1] == 1:
            return self.samples[index, :, :], self.targets[index, :, :]
        else:
            if index < self.data.shape[0] - self.window_size:
                return self.samples[index:index+self.window_size, :], self.targets[index:index+self.window_size, :]
            else:
                return self.samples[-self.window_size:, :], self.targets[-self.window_size:, :]
